tao_compiler/mlir/disc/transforms/conv_rewriter.cc

```diff
@@ -53,19 +54,75 @@ struct ConvParams {
   SmallVector<int64_t> expectedInputLayout;
   SmallVector<int64_t> expectedFilterLayout;
   SmallVector<int64_t> expectedOutputLayout;
-  mhlo::DynamicConvOp conv;
+  T conv;
 };
 
-LogicalResult extractConvParams(mhlo::DynamicConvOp op, ConvParams& params) {
+LogicalResult extractConvParams(mhlo::DynamicConvOp op,
+                                ConvParams<mhlo::DynamicConvOp>& params) {
   params.conv = op;
   params.input = op.lhs();
   params.filter = op.rhs();
   params.output = op.getResult();
 
-  auto inputTy = params.input.getType().dyn_cast<RankedTensorType>();
-  auto filterTy = params.filter.getType().dyn_cast<RankedTensorType>();
-  auto paddingTy = op.d_padding().getType().dyn_cast<RankedTensorType>();
-  auto outputTy = params.output.getType().dyn_cast<RankedTensorType>();
+  auto inputTy = params.input.getType().template dyn_cast<RankedTensorType>();
+  auto filterTy = params.filter.getType().template dyn_cast<RankedTensorType>();
+  auto paddingTy =
+      op.d_padding().getType().template dyn_cast<RankedTensorType>();
+  auto outputTy = params.output.getType().template dyn_cast<RankedTensorType>();
+
+  if (!inputTy || !filterTy || !paddingTy || !outputTy) {
+    return op.emitOpError() << "operands must be ranked type";
+  }
+
+  int rank = filterTy.getRank();
```

Comment:
nit: better to factor out line 77-108. They are the same for `DynamicConvOp` and `QuantizedDynamicConvOp `

Reply:
have refactored